In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from matplotlib import pyplot as plt

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/the-movies-dataset/ratings.csv
/kaggle/input/the-movies-dataset/links_small.csv
/kaggle/input/the-movies-dataset/credits.csv
/kaggle/input/the-movies-dataset/keywords.csv
/kaggle/input/the-movies-dataset/movies_metadata.csv
/kaggle/input/the-movies-dataset/ratings_small.csv
/kaggle/input/the-movies-dataset/links.csv


## Loading the Data

In [2]:
# With so many movies coming out, I wanted to know who which movie had the highest rating and what genre is it in. I have a slight passion with movies so let's find the highest rated movie shall we.
movieData = pd.read_csv('/kaggle/input/the-movies-dataset/movies_metadata.csv', dtype='unicode')
movieData

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                                  genres  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                         [{'id': 35, 'name': 'Comedy'}]   
...                                                  ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462                      [{'id': 18, 'name': 'Drama'}]   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464                                                 []   
45465                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45461                fa                      رگ خواب   
45462                tl          Siglo ng Pagluluwal   
45463                en                     Betrayal   
45464                en          Satana likuyushchiy   
45465                en                     Queerama   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2      A family wedding reignites the ancient feud be...  ...   1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4      Just when George Banks has recovered from his ...  ...   1995-02-10   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisati

## Cleaning the data

in here we will only get the things we need to avoid confusion

In [3]:
# For the Movie Dataset I will only take id,titles and genres.
movieData.dtypes


adult                    object
belongs_to_collection    object
budget                   object
genres                   object
homepage                 object
id                       object
imdb_id                  object
original_language        object
original_title           object
overview                 object
popularity               object
poster_path              object
production_companies     object
production_countries     object
release_date             object
revenue                  object
runtime                  object
spoken_languages         object
status                   object
tagline                  object
title                    object
video                    object
vote_average             object
vote_count               object
dtype: object

In [4]:
selected_columns = ['id', 'title', 'genres','vote_average','vote_count', 'revenue']
new_df = movieData[selected_columns]
new_df.dropna()


id                        title  \
0         862                    Toy Story   
1        8844                      Jumanji   
2       15602             Grumpier Old Men   
3       31357            Waiting to Exhale   
4       11862  Father of the Bride Part II   
...       ...                          ...   
45461  439050                       Subdue   
45462  111109          Century of Birthing   
45463   67758                     Betrayal   
45464  227506             Satan Triumphant   
45465  461257                     Queerama   

                                                  genres vote_average  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...          7.7   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...          6.9   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...          6.5   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...          6.1   
4                         [{'id': 35, 'name': 'Comedy'}]          5.7   
...                                                  ...          ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...          4.0   
45462                      [{'id': 18, 'name': 'Drama'}]          9.0   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...          3.8   
45464                                                 []          0.0   
45465                                                 []          0.0   

      vote_count    revenue  
0           5415  373554033  
1           2413  262797249  
2             92          0  
3             34   81452156  
4            173   76578911  
...          ...        ...  
45461          1          0  
45462          3          0  
45463          6          0  
45464          0          0  
45465          0          0  

[45460 rows x 6 columns]

In [5]:
#let's be more specific
new_df.loc[15480]

id                                                          27205
title                                                   Inception
genres          [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...
vote_average                                                  8.1
vote_count                                                  14075
revenue                                                 825532764
Name: 15480, dtype: object

In [6]:
# finding a movie by it's title
new_df.loc[movieData['title'] == 'Toy Story']

id      title                                             genres  \
0  862  Toy Story  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   

  vote_average vote_count    revenue  
0          7.7       5415  373554033

## Consideration 
### How Can we considere a vote_count is viable, we can see that there are movies with only 1 vote count but have a rating of 10.0 as a vote average.
we need to find the highest value of vote_count and provide a significant level that has a good precision and margin of error levels. in this case we would set it has half of the vote_count to still have a significance level to consider

### Factors to Consider
* Sample Size
* Genre differentials
* Vote Count differentials
* Fine Tuning



In [7]:
#### finding the highest vote count
largestVote = new_df.sort_values(by='vote_count',ascending = False).head(5)
largestVote

id                        title  \
3040    10137                Stuart Little   
3039     2277             Bicentennial Man   
5347      818  Austin Powers in Goldmember   
23224  172385                        Rio 2   
12891   13310         Let the Right One In   

                                                  genres vote_average  \
3040   [{'id': 16, 'name': 'Animation'}, {'id': 14, '...          5.8   
3039   [{'id': 35, 'name': 'Comedy'}, {'id': 878, 'na...          6.9   
5347   [{'id': 35, 'name': 'Comedy'}, {'id': 80, 'nam...          5.9   
23224  [{'id': 16, 'name': 'Animation'}, {'id': 12, '...          6.3   
12891  [{'id': 27, 'name': 'Horror'}, {'id': 18, 'nam...          7.5   

      vote_count    revenue  
3040         998  300135367  
3039         998   93700000  
5347         997  296655431  
23224        997  500188435  
12891        997   10785801

We see that **Inception** has the highest vote_count of all

# Test 1

In [8]:
result1 = new_df.sort_values(by='vote_average',ascending = False).head(5)
result1

id                                              title  \
20395  123648                                           The Land   
19238  320917                                            Oranges   
41232  276833  Quiet Riot: Well Now You're Here, There's No W...   
29494  265330                                           Bad Boys   
41710  380013                               Nobody Walks in L.A.   

                                                  genres vote_average  \
20395                      [{'id': 18, 'name': 'Drama'}]          9.8   
19238  [{'id': 10749, 'name': 'Romance'}, {'id': 18, ...          9.6   
41232  [{'id': 10402, 'name': 'Music'}, {'id': 99, 'n...          9.5   
29494                [{'id': 99, 'name': 'Documentary'}]          9.5   
41710  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...          9.5   

      vote_count revenue  
20395          4       0  
19238          5       0  
41232          2       0  
29494          4       0  
41710          2       0

# Result 1
We can see that the highest rated movie of all time is a movie called **The Land**, the problem with this is that the number of people who voted is 4 and we have atleast 14,000 as the highest vote counts. 

We would need to further provide constraints to get a better value

# Test 2

In [9]:
# since we have the largest number of vote we can use that as the highest total value and half as the lowest value
result2 = new_df.loc[(pd.to_numeric(movieData['vote_average']) < 10 ) & (pd.to_numeric(movieData['vote_count']) > 7037)]
result2.sort_values(by='vote_average', ascending=False).head(3)

id                     title  \
314    278  The Shawshank Redemption   
292    680              Pulp Fiction   
12481  155           The Dark Knight   

                                                  genres vote_average  \
314    [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...          8.5   
292    [{'id': 53, 'name': 'Thriller'}, {'id': 80, 'n...          8.3   
12481  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...          8.3   

      vote_count     revenue  
314         8358    28341469  
292         8670   213928762  
12481      12269  1004558444

# Result 2
The Top highest rated movie of all time is **The Shawshank Redemption** followed by **Pulp Fiction** and **The Dark Knight**.



# Test 3
Improving Level of Significance, Providing a tighter range to see improvement of the level of significance

In [10]:
results3 = new_df.loc[(pd.to_numeric(movieData['vote_average']) < 10 ) & (pd.to_numeric(movieData['vote_count']) > 10000)]
results3.sort_values(by='vote_average', ascending=False).head(3)

id            title  \
12481     155  The Dark Knight   
15480   27205        Inception   
22879  157336     Interstellar   

                                                  genres vote_average  \
12481  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...          8.3   
15480  [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...          8.1   
22879  [{'id': 12, 'name': 'Adventure'}, {'id': 18, '...          8.1   

      vote_count     revenue  
12481      12269  1004558444  
15480      14075   825532764  
22879      11187   675120017

# Result 3

The results shows that the highest rated movie with a higher value of vote count is **The Dark Knight**. we can consider this in a category of a high grossing rated movie because of the amount people who voted but there are still movies highly rated than this. We can also see that **Inception** has the highest vote count among all movies.

# Test 4
Let's find the highest rated movie with more than 1 billion revenue

In [11]:
results4 = new_df.loc[(pd.to_numeric(movieData['vote_average']) < 10 ) & (pd.to_numeric(movieData['vote_count']) > 10000) & (pd.to_numeric(movieData['revenue']) > 1000000000)]
results4.sort_values(by= 'vote_average', ascending = False).head(5)

id            title  \
12481    155  The Dark Knight   
17818  24428     The Avengers   
14551  19995           Avatar   

                                                  genres vote_average  \
12481  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...          8.3   
17818  [{'id': 878, 'name': 'Science Fiction'}, {'id'...          7.4   
14551  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...          7.2   

      vote_count     revenue  
12481      12269  1004558444  
17818      12000  1519557910  
14551      12114  2787965087

# Result 4

This shows us that the highest rated movie of all time and has atleast over 1 billion revenue is **The Dark Knight**.

# Test

* Test 1 - Highest rated movie of all time 

Result -  **The Land**

Issue - It only has 4 votes and has low confidence level since in a sample size of 14,000 it only takes about 0.0003% of the size.
    
* Test 2 - Highest rated movie of all time with atleast 7,000 votes

Result - **Shawshank Redemption** with 14,075 votes

Issue - It may have the highest numberof votes and a significant size but we still can not exclude the fact that it only has about 60% of the total vote count

* Test 3 - Highest rated movie of all time with atleast 10,000 votes

Result - **The Dark Knight** with 12,269 votes which is about 87% of the vote , we also see that **Inception** has the highest vote of all time

* Test 4 -  Highest rated movie of all time with atleast 10,000 votes and atleast 1 billion in revenue.

Result - **The Dark Knight** with 12,269 votes and also $1,004,558,444 in revenue


# Conculsion

In conclusion, the tests conducted to determine the highest rated movie of all time revealed several results. The initial highest-rated movie, **"The Land,"** was found to lack statistical significance due to its low number of votes and limited representation in the overall sample. Moving forward, tests with higher thresholds for the number of votes led to more robust conclusions.

**"The Shawshank Redemption"** emerged as the highest-rated movie with over 7,000 votes, but the issue of it representing only 60% of the total vote count raised concerns. A more refined criterion of at least 10,000 votes highlighted **"The Dark Knight"** as the top contender, comprising 87% of the votes and surpassing even the highest voted movie "**Inception."**

Further refinement, incorporating both high votes and substantial revenue, solidified **"The Dark Knight"** as the conclusive result, boasting 12,269 votes and an impressive $1,004,558,444 in revenue. This comprehensive approach ensures a more reliable determination of the highest rated movie, considering both audience appreciation and commercial success.
